File: PI_4eme-Template/Platforme/src/app/core/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { Notification, NotificationCount } from '../models/notification.model';

@Injectable({ providedIn: 'root' })
export class NotificationService {
  private readonly apiUrl = `${environment.competitionsApiUrl}/notifications`;
  
  private unreadCountSubject = new BehaviorSubject<number>(0);
  unreadCount$ = this.unreadCountSubject.asObservable();

  private notificationsSubject = new BehaviorSubject<Notification[]>([]);
  notifications$ = this.notificationsSubject.asObservable();

  constructor(private http: HttpClient) {}

  // Récupérer toutes les notifications
  getUserNotifications(userId: number): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${this.apiUrl}/user/${userId}`).pipe(
      tap(notifications => this.notificationsSubject.next(notifications))
    );
  }

  // Récupérer les notifications non lues
  getUnreadNotifications(userId: number): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${this.apiUrl}/user/${userId}/unread`);
  }

  // Compter les notifications non lues
  getUnreadCount(userId: number): Observable<NotificationCount> {
    return this.http.get<NotificationCount>(`${this.apiUrl}/user/${userId}/unread/count`).pipe(
      tap(result => this.unreadCountSubject.next(result.count))
    );
  }

  // Marquer comme lue
  markAsRead(notificationId: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${notificationId}/read`, {});
  }

  // Marquer toutes comme lues
  markAllAsRead(userId: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/user/${userId}/read-all`, {}).pipe(
      tap(() => this.unreadCountSubject.next(0))
    );
  }

  // Supprimer une notification
  deleteNotification(notificationId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${notificationId}`);
  }

  // Envoyer une annonce (FORMATEUR)
  sendAnnouncement(competitionId: number | null, title: string, message: string, userIds: number[]): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/announcement`, {
      competitionId,
      title,
      message,
      userIds
    });
  }

  // Ajouter une notification en temps réel (appelé par WebSocket)
  addNotification(notification: Notification): void {
    const current = this.notificationsSubject.value;
    this.notificationsSubject.next([notification, ...current]);
    this.unreadCountSubject.next(this.unreadCountSubject.value + 1);
  }
}

File: Professional Events/src/main/java/org/example/professional_events/service/NotificationService.java
package org.example.professional_events.service;

import org.example.professional_events.dto.NotificationDTO;
import org.example.professional_events.entity.Notification;
import org.example.professional_events.repository.NotificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class NotificationService {

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private SimpMessagingTemplate messagingTemplate;

    // Envoyer une notification à un utilisateur
    public NotificationDTO sendNotification(Long userId, String type, String title, String message, 
                                           Long competitionId, Long relatedId) {
        Notification notification = new Notification();
        notification.setUserId(userId);
        notification.setType(type);
        notification.setTitle(title);
        notification.setMessage(message);
        notification.setCompetitionId(competitionId);
        notification.setRelatedId(relatedId);
        
        notification = notificationRepository.save(notification);
        
        // Envoyer via WebSocket en temps réel
        NotificationDTO dto = convertToDTO(notification);
        messagingTemplate.convertAndSendToUser(
            userId.toString(), 
            "/queue/notifications", 
            dto
        );
        
        return dto;
    }

    // Envoyer une annonce à tous les participants d'une compétition
    public void sendAnnouncementToCompetition(Long competitionId, String title, String message, List<Long> userIds) {
        for (Long userId : userIds) {
            sendNotification(userId, "ANNOUNCEMENT", title, message, competitionId, null);
        }
    }

    // Récupérer toutes les notifications d'un utilisateur
    public List<NotificationDTO> getUserNotifications(Long userId) {
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(userId)
                .stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Récupérer les notifications non lues
    public List<NotificationDTO> getUnreadNotifications(Long userId) {
        return notificationRepository.findByUserIdAndIsReadOrderByCreatedAtDesc(userId, false)
                .stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Compter les notifications non lues
    public Long countUnreadNotifications(Long userId) {
        return notificationRepository.countByUserIdAndIsRead(userId, false);
    }

    // Marquer une notification comme lue
    public void markAsRead(Long notificationId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Notification not found"));
        notification.setIsRead(true);
        notificationRepository.save(notification);
    }

    // Marquer toutes les notifications comme lues
    public void markAllAsRead(Long userId) {
        List<Notification> notifications = notificationRepository.findByUserIdAndIsReadOrderByCreatedAtDesc(userId, false);
        notifications.forEach(n -> n.setIsRead(true));
        notificationRepository.saveAll(notifications);
    }

    // Supprimer une notification
    public void deleteNotification(Long notificationId) {
        notificationRepository.deleteById(notificationId);
    }

    // Convertir entité en DTO
    private NotificationDTO convertToDTO(Notification notification) {
        return new NotificationDTO(
            notification.getIdNotification(),
            notification.getUserId(),
            notification.getType(),
            notification.getTitle(),
            notification.getMessage(),
            notification.getIsRead(),
            notification.getCreatedAt(),
            notification.getCompetitionId(),
            notification.getRelatedId()
        );
    }
}
